### DISCUSSION TITLE
MS Azure Cost Management Plugin has currency type set default as "KRW"

### DISCUSSION THREAD

**Discussion by vivekPrasa** [cloudforet-io/community#141] [Bug Report] [answered]
Hi, 
I'm using azure plugin to enable the azure billing & analysis under cost explorer, and found the the default currency type has been set as KRW "South Korean Won" which is making us difficult to analyse and calculate the cost and budget for Azure specific billings in $USD. 

It would be great if you pls fix this in your code Or make the currency as parameter in configuring the plugin data source and update the Guide pls. 

Below is the link seems where currency has been set default as KRW. 

Link: https://github.com/ImMin5/plugin-azure-cost-mgmt-cost-datasource/blob/master/src/cloudforet/cost_analysis/model/data_source_model.py#L56

Thanks ! 
![2](https://github.com/cloudforet-io/cloudforet/assets/128474463/551d3a70-1c20-4a30-8692-60c806c8bc08)
![3](https://github.com/cloudforet-io/cloudforet/assets/128474463/c8a23426-a2de-402e-bb87-13ffe4768196)

**Comment by ImMin5** [comment]
Hi, @vivekPrasa 
Thanks for your report.

I will fix this issue ASAP 👍

**Reply by vivekPrasa** [reply]
Thanks for your changes, I updated the plugin with currency as USD but in Spaceone UI, I still able to see currency as KRW

-----------------------------------------------------------------------------------------------------------------------
# spacectl exec update_plugin cost_analysis.DataSource -p data_source_id=ds-<xxxxxxxxxxxx>-p domain_id=domain-<xxxxxxxxxxxx>-j '{"options": {"currency": "USD"}}'
---
created_at: '2024-01-31T14:18:49.223Z'
data_source_id: ds-<xxxxxxxxxxxx>
data_source_type: EXTERNAL
domain_id: domain-<xxxxxxxxxxxx>
name: Azure Cost Explorer Data Source
plugin_info:
  metadata:
    **currency: KRW**
    data_source_rules:
    - actions:
        match_service_account:
          source: additional_info.Subscription Id
          target: data.subscription_id
      conditions: []
      conditions_policy: ALWAYS
      name: match_service_account
      options:
        stop_processing: true
      tags: {}
    supported_secret_types:
    - MANUAL
  options:
    **currency: USD**
  plugin_id: plugin-azure-cost-mgmt-cost-datasource
  secret_id: secret-<XXXXXXXXXXXX>
  upgrade_mode: AUTO
  version: 1.1.5
provider: azure
secret_type: MANUAL
state: ENABLED
tags: {}
template: {}

-----------------------------------------------------------
<img width="1189" alt="1" src="https://github.com/cloudforet-io/cloudforet/assets/128474463/bf32dbc1-933b-4574-a174-e8e866d03099">

**Reply by vivekPrasa** [reply]
@ImMin5  need additional help over spacectl cli tool, can you pls tell how I can delete already configured data source plugin using spacectl cli tool ?

**Reply by choonho** [reply]
@vivekPrasa the delete datasource is "deregister" API, https://cloudforet-io.github.io/api-doc-1-12/cost_analysis/v1/DataSource/#deregister

spacectl exec deregister cost_analysis.DataSource -p data_source_id={Your Data Source ID} -p cascade_delete_cost=true

**Comment by choonho** [comment]
@ImMin5 One of my guess is updating option does not apply to plugin which was already deployed with previous configuration.

**Reply by vivekPrasa** [reply]
@choonho @ImMin5 I deregistered the azure plugin (which still has currency showing as KRW even after plugin update for currency USD) and re-registered the plugin from scratch and it still shows default currency as KRW.  

<img width="1166" alt="image" src="https://github.com/cloudforet-io/cloudforet/assets/128474463/504efcb8-ec5d-494b-aa40-450f97842874">




Another issue which I'm experiencing with Azure plugin is that I'm unable to sync the the data with my azure account. I have passed the billing account details along with Client ID, Client Secret and Tenant Id and reverified the permissions of the account to have billing access multiple times but still it shows me access forbidden error. Kindly help and advise. Thanks 

<img width="795" alt="image" src="https://github.com/cloudforet-io/cloudforet/assets/128474463/8423e591-e3fe-46b9-973e-e91259df9900">

**Reply by vivekPrasa** [reply]
@ImMin5 just checking if you got anything on above request.

**Reply by vivekPrasa** [reply]
Hi @ImMin5 @choonho  Still stuck with above issues, could you pls help here ? Thanks !

**Reply by choonho** [reply, marked as answer]
@vivekPrasa I think Minho already fixed the code from the commit.

https://github.com/cloudforet-io/plugin-azure-cost-mgmt-cost-datasource/commit/45168b025e1952569b992234763800bc9cd474c2

The reason what your environment is not working is that you are using default configuration for plugins which is for developer.

My guess is
## Upgrade helm value file.

* Add repository section
~~~
repository:
  enabled: true
  application_grpc:
    MANAGED_PLUGIN_IMAGE_PREFIX: cloudforet
~~~~

* helm upgrade
* You didn't need to delete all pods. repository POD will be re-created.
~~~
helm upgrade cloudforet cloudforet/spaceone -n spaceone -f values.yaml
~~~

* De-register previous data-source
~~~
spacectl exec deregister cost_analysis.DataSource -p data_source_id={Your Data Source ID}
~~~

* Register data-source with options.currency=USD

If you have a problem again, Join following Google Meet.

https://calendar.google.com/calendar/u/0?cid=MTdlYzU2MjFjNDlhN2Y3ZGFlMzc1NWJhM2EwNjhkZmNhNDA2NTY0OGMxYTM4ZGZmMWE3MWViZjlmMGIwZDU1NUBncm91cC5jYWxlbmRhci5nb29nbGUuY29t

I will invite Minho,

CET: 20 Feb. 15:00 pm
KST: 20 Feb. 23:00 pm
 
Azure Data Source
https://meet.google.com/nws-fvnv-crm

**Comment by choonho** [comment]
@ImMin5 to change currency, you have to upgrade plugin.

When client registers datasource in options.currency, you have to return in metadata.currency

The currency should be returned in ***plugin_info.metadat.currency***

Current response looks like

~~~
root@spacectl-867bbf45f4-dxdpg:~# spacectl exec register cost_analysis.DataSource -f azure_data_source.yaml
---
created_at: '2024-02-13T13:19:46.451Z'
data_source_id: ds-2d88d5d16e98
data_source_type: EXTERNAL
domain_id: domain-a794f8bd029d
name: Azure
plugin_info:
  metadata:
    currency: KRW
    data_source_rules:
    - actions:
        match_service_account:
          source: additional_info.Subscription Id
          target: data.subscription_id
      conditions: []
      conditions_policy: ALWAYS
      data_source_id: ds-2d88d5d16e98
      domain_id: domain-a794f8bd029d
      name: match_service_account
      options:
        stop_processing: true
      rule_type: MANAGED
      tags: {}
    supported_secret_types:
    - MANUAL
  options:
    collect_resource_id: true
    currency: USD
  plugin_id: plugin-azure-cost-mgmt-cost-datasource
  upgrade_mode: AUTO
  version: 1.1.5
provider: azure
secret_type: USE_SERVICE_ACCOUNT_SECRET
state: ENABLED
tags: {}
template: {}
~~~

plugin_info.options.currency: USD
plugin_info.metadata.currency: KRW (I think that frontend uses plugin_info.metadata.currency)

**Comment by choonho** [comment]
Actually currency is just frontend value, the real data depends on your setting, like USD, EURO or SEK.
From the log, the first issue is permission for billing API. You have to update permission like

https://github.com/ImMin5/plugin-azure-cost-mgmt-cost-datasource/blob/master/docs/en/GUIDE.md#register-the-datasource-with-azure-credentials

https://learn.microsoft.com/en-us/azure/cost-management-billing/automate/cost-management-api-permissions#assign-service-principal-access-to-azure-resource-manager-apis

**Reply by ImMin5** [reply]
If you want to register EA Agreement billing Account this [official document](https://learn.microsoft.com/en-us/azure/cost-management-billing/manage/assign-roles-azure-service-principals) will help you. or  
If you want to register except EA Agreement please follow this [official document](https://learn.microsoft.com/en-us/azure/cost-management-billing/manage/understand-mca-roles#manage-billing-roles-in-the-azure-portal)

thanks :)
